pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -585,6 +595,118 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
 	return returnErr
 }
 
+func (grb *globalRoleBindingLifecycle) reconcileNamespacedRoleBindings(globalRoleBinding *v3.GlobalRoleBinding) error {
+	var returnError error
+	gr, err := grb.grLister.Get("", globalRoleBinding.GlobalRoleName)
+	if err != nil {
+		return fmt.Errorf("unable to get globalRole %s: %w", globalRoleBinding.GlobalRoleName, err)
+	}
+
+	// get all the roleBindings claiming to be owned by this GRB and remove any that shouldn't exist
+	r, _ := labels.NewRequirement(grbOwnerLabel, selection.Equals, []string{globalRoleBinding.Name})
+	rbs, err := grb.roleBindingLister.List("", labels.NewSelector().Add(*r))
+	if err != nil {
+		returnError = multierror.Append(returnError,
+			errors.Wrapf(err, "couldn't list roleBindings with label %s : %s", grbOwnerLabel, globalRoleBinding.Name))
+	} else {
+		err = grb.purgeInvalidRoleBindings(rbs, gr.NamespacedRules)
+		if err != nil {
+			returnError = multierror.Append(returnError, err)
+		}
+	}
+
+	for ns := range gr.NamespacedRules {
+		_, err := grb.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+
+		// Check if the role we're binding to exists
+		roleName := gr.Name + "-" + ns
+		_, err = grb.roleLister.Get(ns, roleName)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get role %s", roleName))
+			continue
+		}
+
+		subject := rbac.GetGRBSubject(globalRoleBinding)
+		rbName := globalRoleBinding.Name + "-" + ns
+		roleBinding, err := grb.roleBindingLister.Get(ns, rbName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+
+			newRoleBinding := &v1.RoleBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      rbName,
+					Namespace: ns,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: globalRoleBinding.APIVersion,
+							Kind:       globalRoleBinding.Kind,
+							UID:        globalRoleBinding.UID,
+							Name:       globalRoleBinding.Name,
+						},
+					},
+					Labels: map[string]string{
+						grbOwnerLabel: globalRoleBinding.Name,
+					},
+				},
+				RoleRef: v1.RoleRef{
+					Name: roleName,
+					Kind: "Role",
+				},
+				Subjects: []v1.Subject{subject},
+			}
+			_, err := grb.roleBindings.Create(newRoleBinding)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
I think in the case that this already exists, we need to do an integrity check on the existing role, correct? In this case, if we have a corrupt binding (i.e. referring to an invalid role) in place, we won't correct that for a long time because of this logic.